File: bindings/julia/src/SafePath.jl
# SPDX-License-Identifier: Apache-2.0
# SPDX-FileCopyrightText: 2025 Hyperpolymath

"""
    SafePath

Safe filesystem path operations with traversal attack prevention.
"""
module SafePath

export has_traversal, is_safe, sanitize_filename, safe_join

"""
    has_traversal(path::AbstractString) -> Bool

Check if a path contains directory traversal sequences.
"""
function has_traversal(path::AbstractString)::Bool
    contains(path, "..") || contains(path, "~")
end

"""
    is_safe(path::AbstractString) -> Bool

Check if a path is safe (no traversal attacks).
"""
function is_safe(path::AbstractString)::Bool
    !has_traversal(path)
end

"""
    sanitize_filename(filename::AbstractString) -> String

Sanitize a filename by removing dangerous characters.
"""
function sanitize_filename(filename::AbstractString)::String
    filename |>
    x -> replace(x, ".." => "_") |>
    x -> replace(x, "/" => "_") |>
    x -> replace(x, "\\" => "_") |>
    x -> replace(x, "<" => "_") |>
    x -> replace(x, ">" => "_") |>
    x -> replace(x, ":" => "_") |>
    x -> replace(x, "\"" => "_") |>
    x -> replace(x, "|" => "_") |>
    x -> replace(x, "?" => "_") |>
    x -> replace(x, "*" => "_") |>
    x -> replace(x, "\0" => "_")
end

"""
    safe_join(base::AbstractString, parts::AbstractVector{<:AbstractString}) -> Union{String, Nothing}

Safely join path components, rejecting traversal attempts.
Returns `nothing` if any part contains traversal sequences.
"""
function safe_join(base::AbstractString, parts::AbstractVector{<:AbstractString})::Union{String, Nothing}
    any(has_traversal, parts) && return nothing

    sanitized = map(sanitize_filename, parts)
    path = base
    for part in sanitized
        path = endswith(path, "/") ? path * part : path * "/" * part
    end
    path
end

"""
    safe_join(base::AbstractString, parts::AbstractString...) -> Union{String, Nothing}

Variadic version of safe_join.
"""
function safe_join(base::AbstractString, parts::AbstractString...)::Union{String, Nothing}
    safe_join(base, collect(parts))
end

end # module

File: bindings/ocaml/lib/safe_path.ml
(* SPDX-License-Identifier: Apache-2.0 *)
(* SPDX-FileCopyrightText: 2025 Hyperpolymath *)

(** Safe filesystem path operations with traversal attack prevention. *)

let has_traversal path =
  String.length path >= 2 &&
  (let rec check i =
    if i > String.length path - 2 then false
    else if path.[i] = '.' && path.[i+1] = '.' then true
    else check (i + 1)
  in check 0) ||
  String.contains path '~'

let is_safe path = not (has_traversal path)

let sanitize_filename filename =
  let buf = Buffer.create (String.length filename) in
  let prev_dot = ref false in
  String.iter (fun c ->
    match c with
    | '.' when !prev_dot ->
        Buffer.add_char buf '_';
        prev_dot := false
    | '.' ->
        prev_dot := true;
        Buffer.add_char buf '.'
    | '/' | '\\' | '<' | '>' | ':' | '"' | '|' | '?' | '*' | '\000' ->
        if !prev_dot then prev_dot := false;
        Buffer.add_char buf '_'
    | c ->
        if !prev_dot then prev_dot := false;
        Buffer.add_char buf c
  ) filename;
  (* Handle trailing .. *)
  let result = Buffer.contents buf in
  let len = String.length result in
  if len >= 2 && result.[len-2] = '.' && result.[len-1] = '.' then
    String.sub result 0 (len - 2) ^ "__"
  else result

let safe_join base parts =
  if List.exists has_traversal parts then None
  else
    let sanitized = List.map sanitize_filename parts in
    Some (List.fold_left Filename.concat base sanitized)
